src/node.cc

```diff
@@ -2703,90 +2720,160 @@ struct DLib {
     uv_dlclose(&lib_);
   }
 #endif  // !__POSIX__
+
+  DLib(const DLib& other) = delete;
+  DLib(DLib&& other) = delete;
+  DLib& operator=(const DLib& other) = delete;
+  DLib& operator=(DLib&& other) = delete;
+
+  void AddEnvironment(Environment* env) {
+    if (users_.count(env) > 0) return;
+    users_.insert(env);
+    if (env->is_main_thread()) return;
+    struct cleanup_hook_data {
+      std::shared_ptr<DLib> info;
+      Environment* env;
+    };
+    env->AddCleanupHook([](void* arg) {
+      Mutex::ScopedLock lock(dlib_mutex);
+      cleanup_hook_data* cbdata = static_cast<cleanup_hook_data*>(arg);
+      std::shared_ptr<DLib> info = cbdata->info;
+      info->users_.erase(cbdata->env);
+      delete cbdata;
+      if (info->users_.empty()) {
+        std::vector<std::string> filenames;
+
+        for (const auto& entry : dlopen_cache) {
+          if (entry.second == info)
+            filenames.push_back(entry.first);
+        }
+        for (const std::string& filename : filenames)
+          dlopen_cache.erase(filename);
+
+        handle_to_dlib.erase(info->handle_);
+      }
+    }, static_cast<void*>(new cleanup_hook_data { shared_from_this(), env }));
+  }
 };
 
+}  // anonymous namespace
+
 // DLOpen is process.dlopen(module, filename, flags).
 // Used to load 'module.node' dynamically shared objects.
-//
-// FIXME(bnoordhuis) Not multi-context ready. TBD how to resolve the conflict
-// when two contexts try to load the same shared object. Maybe have a shadow
-// cache that's a plain C list or hash table that's shared across contexts?
 static void DLOpen(const FunctionCallbackInfo<Value>& args) {
   Environment* env = Environment::GetCurrent(args);
+  node_module* mp;
+  std::shared_ptr<DLib> dlib;
+  Local<Object> module = args[0]->ToObject(env->isolate());
 
-  CHECK_EQ(modpending, nullptr);
+  do {
```

Comment:
Is there any particular reason to use the `do { ... } while (false);` pattern instead of a plain `{ ... }` scope here?

Reply:
Yeah, the fact that you can `break;` out of it if you discover that you don’t want to do most of the work here. The alternative would be having quite a sizeable block of code with 2 levels of extra indentation, which would also work

Reply:
Oh, right. I just noticed the loop when I have opened the PR again, without really looking into its body this time, so I didn't notice the `break` statements, heh. Thanks for clarifying!